26-dracula-ant/technical_specification.md

```diff
@@ -55,38 +57,59 @@ Deletion of work packages should be logged or implemented as deactivation (immed
 
 ### Work order token
 
 This is a JWT with a very short time to live (maximum 30 seconds). It is signed by the Work Package Service and then encrypted with the `public_key` of the user, similar to the work package access token. It contains the following claims:
 
 - `type`: `download` or `upload`
 - `file_id`: the ID of the file that shall be downloaded or uploaded
+- `file_ext`: the file extension of the requested filed including the dot
 - `user_id`: the internal ID of the user
-- `public_key`: the public key stored in the work package
+- `public_key`: the public key of the user stored in the work package
 - `full_user_name`: the full name of the user (with academic title)
 - `email`: the email address of the user
 
 The user name and email are part of the token so that the upload/download controllers can log this information and notify the users.
 
-The work order tokens are signed by the Work Package Service. A separate key pair (different from the keys used for the internal access tokens) should be created for this purpose. The public key must be also communicated to the download/upload controller services so that they can validate the work order token.
+The work order tokens are signed by the Work Package Service. A separate key pair (different from the keys used for the internal access tokens) is created for this purpose, stored by the HashiCorp vault and injected into the service via configuration. The public key must be provided to the download/upload controller services so that they can validate the work order token, the private key must only be made available to the Work Package Service.
 
 ### Dataset access checks
 
 #### Access checks by the Work Package Service
 
 To allow the Work Package Service to check whether a given user has *download* access to a given dataset, the claims repository provides the following *internal* endpoint:
 
-- `GET /datasets/{dataset_id}/users/{user_id}/download-access`
+- `GET /download-access/users/{user_id}/datasets`
+  - authorization: only internal from download controller (via Istio)
+  - returns a list of all dataset IDs that can be downloaded by the user
+
+- `GET /download-access/users/{user_id}/datasets/{dataset_id}`
   - authorization: only internal from download controller (via Istio)
   - returns `true` or `false` as a scalar resource
 
-In order to facilitate authorization, the path of this endpoint starts with `datasets` and not with `users` which is already used by other endpoints fo the claims repository.
+*TODO:* Instead of true/false, the latter endpoint could also return null or the date when the access expires (null could be confused with unlimited access though). This could be used by the Work Package Service to limit the expiration date of the tokens, even though the Work Package Service re-checks access on every operation anyway.
```

Comment:
I think this is a good idea but maybe not that essential now. Should we do this as part of another work package?

Reply:
What would the advantage of using a date be compared to just checking, i.e. what's the context here?

Reply:
@mephenor The advantage would be that the WPS could use this as upper limit for the expiration date of the work package, and that the expiration date could also be communicated to the users in the frontend.